### PR TITLE
Made changes so that tests pass

### PR DESCRIPTION
**Pull request by Iainmon**
This PR makes changes so that the tests don't fail.

- [x] Run tests

### CODE REVIEW

File: test/studies/ml/lib/Chai.chpl [hunk 1]
```diff
@@ -137,32 +137,32 @@ module Chai {
...
 
-        proc forwardProp(x: Tensor(?rank)): Tensor(rank) {
+        proc ref forwardProp(x: Tensor(?rank)): Tensor(rank) {
             return tn.sigmoid(x);
         }
 
-        proc forwardPropBatch(batch: [] ?tensorType) where isSubtype(tensorType, Tensor) {
-            return tn.sigmoid(batch);
+        proc ref forwardPropBatch(batch: [] ?tensorType) where isSubtype(tensorType, Tensor) {
+            return forwardProp(batch);
```
Comment (e-kayrakli) [inline test/studies/ml/lib/Chai.chpl:145]: You also changed the function you called here. Is that OK?

File: test/studies/ml/lib/Chai.chpl [hunk 2]
```diff
@@ -137,32 +137,32 @@ module Chai {
...
-        proc backward(delta: Tensor(?rank),lastInput: Tensor(rank)): Tensor(rank) {
+        proc const ref backward(delta: Tensor(?rank),lastInput: Tensor(rank)): Tensor(rank) {
             return delta * tn.sigmoidPrime(lastInput); 
         }
 
-        proc backwardBatch(deltas: [] ?tensorType1, inputs: [] ?tensorType2) 
+        proc const ref backwardBatch(deltas: [] ?tensorType1, inputs: [] ?tensorType2) 
             where isSubtype(tensorType1, Tensor) && isSubtype(tensorType2, Tensor) {
-            return [(d,i) in zip(deltas,inputs)] backward(d,i);
+            return backward(deltas,inputs);
```
Comment (e-kayrakli) [inline test/studies/ml/lib/Chai.chpl:154]: Same here. So I'm guessing that's intentional

File: test/studies/ml/lib/Chai.chpl [hunk 3]
```diff
@@ -366,23 +361,23 @@ module Chai {
             return grad;
         }
 
-        proc backwardBatch(deltas: [] Tensor(3), inputs: [] Tensor(3)): [] Tensor(3) {
+        proc ref backwardBatch(deltas: [] Tensor(3), inputs: [] Tensor(3)): [] Tensor(3) {
             var newDeltas: [deltas.domain] Tensor(3);
-            forall (delta,inputs,newDelta) in zip(deltas,inputs,newDeltas) {
+            forall (delta,inputs,newDelta) in zip(deltas,inputs,newDeltas) with (ref this) {
```
Comment (e-kayrakli) [inline test/studies/ml/lib/Chai.chpl:366]: This is interesting, I asked whether we should require this in https://github.com/chapel-lang/chapel/issues/23229#issuecomment-1710882522

File: test/studies/ml/lib/Chai.chpl [hunk 4]
```diff
@@ -436,27 +431,27 @@ module Chai {
...
+            return forwardProp(batch);
 
-        proc backward(delta: Tensor(1), input: Tensor(?inRank)): Tensor(inRank) do
+        proc ref backward(delta: Tensor(1), input: Tensor(?inRank)): Tensor(inRank) do
             return delta.reshape(input.domain);
 
-        proc backwardBatch(deltas: [] Tensor(1), inputs: [] ?tensorType): [] tensorType where isSubtype(tensorType,Tensor) do
-            return [(d,i) in zip(deltas,inputs)] backward(d,i);
+        proc ref backwardBatch(deltas: [] Tensor(1), inputs: [] ?tensorType): [] tensorType where isSubtype(tensorType,Tensor) do
+            return backward(deltas,inputs);
```
Comment (e-kayrakli) [inline test/studies/ml/lib/Chai.chpl:444]: Oh, I'm now realizing that this may be just using promotion?

**Review by e-kayrakli — APPROVED** [review]
Not fully aware of failures caused by this, but changes look reasonable.

File: test/studies/ml/lib/Chai.chpl (hunk 1 above)
Comment (Iainmon) [reply test/studies/ml/lib/Chai.chpl:145]: Yes, I was getting new promotion errors.

File: test/studies/ml/lib/Chai.chpl (hunk 2 above)
Comment (Iainmon) [reply test/studies/ml/lib/Chai.chpl:154]: Yes, this seemed to alleviate some of the `ref` errors.

File: test/studies/ml/lib/Chai.chpl (hunk 4 above)
Comment (Iainmon) [reply test/studies/ml/lib/Chai.chpl:444]: Yes

File: test/studies/ml/lib/Chai.chpl (hunk 3 above)
Comment (bradcray) [reply test/studies/ml/lib/Chai.chpl:366]: This seems loosely related to https://github.com/chapel-lang/chapel/issues/19211 and particularly its [most recent comment](https://github.com/chapel-lang/chapel/issues/19211#issuecomment-1236028838).

**Pull request merged**